### ISSUE TITLE
Inquiry Regarding Implementation-Paper Discrepancies in MARS-SQL

### ISSUE THREAD

**Issue by jirishma081** [YangHaolin0526/MARS-SQL#2]
Dear MARS-SQL Authors,

I am currently studying your paper "MARS-SQL: A Multi-Agent Reinforcement Learning Framework for Text-to-SQL" and examining the provided open-source implementation at https://github.com/YangHaolin0526/MARS-SQL. During my analysis, I have identified several significant discrepancies between the paper's claims and the actual code implementation that require clarification.

## 1. Single Model vs. Three Agents Discrepancy

**Finding in Paper (Section 3.1):**
> "Our experimental setup consists of three distinct agents: a Grounding Agent, a Generation Agent, and a Validation Agent."

**Finding in Code:**
The inference script (`Mars-inference/inference.sh`, line 11) shows only one model being used:
```bash
model_path=Yanghl0526/Qwen-SQL-7B-bird_5turns_80step
```

Throughout the entire codebase (`Mars-inference/verl/workers/fsdp_workers.py`), I can only find initialization and usage of a single model that appears to handle all roles through prompt engineering and structured tags (`<think>`, `<sql>`, `<observation>`, `<solution>`).

**Question:** Are there actually three separately trained models as described in the paper? If yes, where are the other two models and why does the inference only use one?

## 2. Training Methodology Mismatch

**Finding in Paper:**
> "The Grounding and Generation Agents were trained using Reinforcement Learning (RL). The Grounding Agent was developed with the Verl framework, using training data prepared with SQLGlot. The Generation Agent utilized a framework adapted from SkyRL."
> "The Validation Agent was trained via full-parameter Supervised Fine-tuning (SFT)"

**Finding in Code:**
The training implementation in `Mars-train/verl/trainer/main_ppo.py` exclusively uses PPO (Proximal Policy Optimization):
- Import statement: `from verl.trainer.ppo.ray_trainer import RayPPOTrainer`
- Configuration files all reference PPO training
- No evidence of GRPO (mentioned in the paper) or separate SFT training for validation

**Question:** Was PPO the actual training method used? How does this relate to the GRPO and SFT methods mentioned in the paper?

## 3. Pre-processed Schema vs. Dynamic Schema Selection

**Critical Finding:**
The paper emphasizes "reasoning-driven schema identification" by the Grounding Agent as a key contribution. However, examining the data files reveals a different reality.

**In `data/bird_test.parquet`:**
The schemas are already pre-processed and pruned:
```python
# Table: frpm
[
(County Name:TEXT, Examples: [Alameda, Alpine, Amador]),
(Enrollment (K-12):REAL, Examples: [1087.0, 395.0, 244.0]),
(Free Meal Count (K-12):REAL, Examples: [565.0, 186.0, 134.0])
]
```

This shows:
- Only relevant tables included (not the full database schema)
- Example values already provided
- Format standardized

**In `Mars-train/verl/trainer/main_ppo.py` (lines 269-328):**
There is extensive commented-out code for dynamic schema building:
```python
# def build_schema_from_names(db_id: str, base_db_path: str, model_output: str) -> str:
#     """
#     The main "glue" function. It parses the model's output to get table/column lists,
#     then uses the helper functions to build the simplified CREATE TABLE statements.
#     """
```

**Questions:**
1. Who or what performed the schema pruning found in the parquet files?
2. If this was done by a model, why is that model not included in the inference pipeline?
3. If this was manual preprocessing, how does this align with the paper's claim of "reasoning-driven schema identification"?
4. Does the reported performance (77.84% on BIRD) include this preprocessing advantage?

## 4. Impact on Benchmarking Fairness

Given that the input data contains pre-selected, relevant schemas rather than full database schemas:
- How does this affect fair comparison with other methods that work with complete schemas?
- Should this preprocessing be considered part of the system's performance or a data preparation step?

## 5. Request for Clarification

To ensure reproducibility and scientific integrity, please provide:

1. **Model Architecture:** Clarification on whether one or three models were actually trained and used
2. **Training Details:** The actual training methods used (PPO vs. GRPO/SFT)
3. **Schema Processing:**
   - The exact method used for schema pruning
   - Whether this was automated or manual
   - The code/model responsible for this if automated
4. **Fair Comparison:** How the preprocessing affects benchmark comparisons

## 6. Code Completeness

Several critical components appear to be missing from the repository:
- Training code for the claimed Grounding Agent
- Training code for the claimed Validation Agent
- Schema preprocessing pipeline
- The other two models (if they exist)

Could you please provide these components or explain their absence?

---

**Disclaimer:** All code analysis above is based on your official public GitHub repository (https://github.com/YangHaolin0526/MARS-SQL), analyzed on November 22, 2025. If there are other versions of the code or if I have missed certain components, please correct me.

I raise these questions in the interest of academic transparency and reproducibility. The Text-to-SQL community would greatly benefit from understanding the actual architecture and methodology behind your impressive results.

Looking forward to your clarification on these matters.

**Comment by ravi07bec** [comment]
+1 to above issue. Also curious to know the BIRD-test set performance since 77.84% is a self-validated dev set claim.
